### PR TITLE
Client Diagnostics plugin

### PR DESCRIPTION
**Pull request by RJ**
Aggregates a few stats while processing packets, and flushes them to bevy's Diagnostics system every second.

I like the [ScreenDiagnosticsPlugin](https://github.com/laundmo/bevy_screen_diagnostics) for quickly displaying various diagnostics. I have it set up like this:

```rust
use bevy_replicon::prelude::ReplicationDiagnostics::*;

app
  .add_plugins(ScreenDiagnosticsPlugin::default())
  .add_plugins(ClientDiagnosticsPlugin)
  .add_systems(Startup, setup_replicon_diagnostics);

// app.add_plugins(ScreenDiagnosticsPlugin::default())
fn setup_replicon_diagnostics(mut diags: ResMut<ScreenDiagnostics>) {
    diags
        .add(
            "r:c/s".to_string(),
            ClientDiagnosticsPlugin::COMPONENT_CHANGES,
        )
        .format(|v| format!("{v:.1}"));

    diags
        .add("r:e/s".to_string(), ClientDiagnosticsPlugin::ENTITY_CHANGES)
        .format(|v| format!("{v:.1}"));

    diags
        .add("r:p/s".to_string(), ClientDiagnosticsPlugin::PACKETS)
        .format(|v| format!("{v:.0}"));

    diags
        .add("r:b/s".to_string(), ClientDiagnosticsPlugin::BYTES)
        .format(|v| format!("{v:.0}"));
}
```

https://github.com/lifescapegame/bevy_replicon/assets/29747/d22445d5-491e-4de2-9085-9abd6b72f798

### CODE REVIEW

File: src/client.rs [hunk 1]
```diff
@@ -55,56 +59,63 @@ impl ClientPlugin {
...
-                            &mut cursor,
-                            world,
-                            &mut entity_map,
-                            &replication_rules,
-                            DiffKind::Change,
-                            tick,
-                        )?;
-                        if cursor.position() == end_pos {
-                            continue;
+                    world.resource_scope(|world, mut stats: Mut<ReplicationStats>| {
```
[the remark this answers is not in the document]
Comment (RJ) [reply src/client.rs:62]: for some reason i didn't expect Option to work with resource_scope, not sure why. will def make it optional.

[the remark this answers is not in the document]
Comment (RJ) [reply src/client.rs:62]: actually i don't think it's possible to resource_scope something Optionally. I can remove the resource the reinsert manually.

**Comment by RJ** [comment]
Made the ReplicationStats resource optional.

still adding the plugin in the ClientPlugin though.

Could add it to the PluginGroup, but it's only really relevant for clientside stats.

File: src/client.rs [hunk 2]
```diff
@@ -16,13 +16,17 @@ use crate::replicon_core::{
...
+use diagnostics::{ReplicationStats, ReplicationStatsPlugin};
+
 pub struct ClientPlugin;
 
 impl Plugin for ClientPlugin {
     fn build(&self, app: &mut App) {
         app.add_plugins((RenetClientPlugin, NetcodeClientPlugin))
             .init_resource::<LastRepliconTick>()
             .init_resource::<ServerEntityMap>()
+            .add_plugins(ReplicationStatsPlugin)
```
Comment (Shatur) [inline src/client.rs:29]: Looks like in Bevy plugins like this aren't enabled by default.
Maybe make this plugin public?
Also looks like Bevy use [`Diagnostics` in names](https://docs.rs/bevy/latest/bevy/diagnostic/index.html). So how about to rename it into `ReplicationDiagnosticsPlugin`?

File: src/client.rs [hunk 3]
```diff
@@ -55,18 +59,26 @@ impl ClientPlugin {
         world.resource_scope(|world, mut client: Mut<RenetClient>| {
             world.resource_scope(|world, mut entity_map: Mut<ServerEntityMap>| {
                 world.resource_scope(|world, replication_rules: Mut<ReplicationRules>| {
+                    let mut stats = world.remove_resource::<ReplicationStats>();
                     while let Some(message) = client.receive_message(REPLICATION_CHANNEL_ID) {
                         let end_pos: u64 = message.len().try_into().unwrap();
                         let mut cursor = Cursor::new(message);
 
+                        if let Some(ref mut rs) = stats {
```
Comment (Shatur) [inline src/client.rs:67]: Let's use `&mut` instead, it's just more common.
Also could you use the same name?
```suggestion
                        if let Some(stats) = &mut stats {
```

File: src/client/diagnostics.rs [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    prelude::*,
+    time::common_conditions::on_timer,
+};
+use std::time::Duration;
+
+/// Collects replication stats during packet processing
+/// flushed to Diagnostics system periodically.
+#[derive(Default, Resource, Debug)]
+pub struct ReplicationStats {
+    // incremented per entity that changes
```
Comment (Shatur) [inline src/client/diagnostics.rs:13]: Use doc comments:
```suggestion
    /// Incremented per entity that changes.
```

File: src/client/diagnostics.rs [hunk 5]
```diff
@@ -0,0 +1,137 @@
+use bevy::diagnostic::DiagnosticId;
+use bevy::{
+    diagnostic::{Diagnostic, Diagnostics, RegisterDiagnostic},
+    prelude::*,
+    time::common_conditions::on_timer,
+};
+use std::time::Duration;
+
+/// Collects replication stats during packet processing
+/// flushed to Diagnostics system periodically.
```
Comment (Shatur) [inline src/client/diagnostics.rs:10]: Official rust do usually styled like this:
```suggestion
/// Replication stats during packet processing.
///
/// Flushed to Diagnostics system periodically.
```

File: src/client/diagnostics.rs [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+    pub mappings: u32,
+    // incrementer per entity despawn
+    pub despawns: u32,
+    // replication packets recvd
+    pub packets: u32,
+    // replication bytes received as packet payload (not transport layer headers, etc)
+    pub bytes: u32,
+}
+
+pub mod replication_diagnostics {
```
Comment (Shatur) [inline src/client/diagnostics.rs:27]: Instead of using a separate module, just put it under the plugin impl.

File: src/client/diagnostics.rs [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+    use super::*;
+    /// How many entities modified per second by replication
+    pub const REPLICATED_ENTITY_CHANGES_DIAGNOSTIC: DiagnosticId =
```
Comment (Shatur) [inline src/client/diagnostics.rs:30]: Let's remove `_DIAGNOTICS` suffix and maybe `REPLICATED_` prefix. Also I would remove comments too, it's obvious from the name. See how it's done in Bevy: https://github.com/bevyengine/bevy/blob/main/crates/bevy_diagnostic/src/frame_time_diagnostics_plugin.rs
```suggestion
    pub const ENTITY_CHANGES: DiagnosticId =
```

File: src/client/diagnostics.rs [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+        .register_diagnostic(Diagnostic::new(
+            REPLICATED_BYTES_DIAGNOSTIC,
+            "bytes per second",
+            REPLICATED_DIAGNOSTIC_HISTORY_LEN,
+        ));
+    }
+}
+
+/// flushes aggregated stats to diagnostics system
+fn write_diagnostics(mut stats: ResMut<ReplicationStats>, mut diagnostics: Diagnostics) {
```
Comment (Shatur) [inline src/client/diagnostics.rs:99]: Fix doc comment and move it under the `impl`.

**Review by Shatur — CHANGES_REQUESTED** [review]
Thank you a lot! Left a few minor suggestions.

**Comment by Shatur** [comment]
And could you please update the changelog too?

File: src/client.rs (hunk 2 above)
Comment (Shatur) [reply src/client.rs]: Actually, I would suggest `ClientDiagnosticsPlugin`. In the future we could have similar thing for server.

**Comment by RJ** [comment]
Have made the suggested changes, thanks 🙂

I kept comments on the DiagnosticIds, because they show up on the docs page for the plugin struct, which is useful.

The `ReplicationStats` resource is public, because it's possible to insert that resource yourself, without using the plugin. Then you are free to aggregate the stats as you like, without needing the Diagnostics system.

File: src/client/diagnostics.rs [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+};
+use std::time::Duration;
+
+/// Replication stats during packet processing.
+///
+/// Flushed to Diagnostics system periodically.
+#[derive(Default, Resource, Debug)]
+pub struct ReplicationStats {
+    /// Incremented per entity that changes.
+    pub entities_changed: u32,
```
Comment (Shatur) [inline src/client/diagnostics.rs:15]: Maybe it worth to use `u64` for all stats?
Looks like Bevy usually use 8 bytes for diagnostic data.

**Review by Shatur — CHANGES_REQUESTED** [review]

File: src/client/diagnostics.rs [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+    prelude::*,
+    time::common_conditions::on_timer,
+};
+use std::time::Duration;
+
+/// Replication stats during packet processing.
+///
+/// Flushed to Diagnostics system periodically.
+#[derive(Default, Resource, Debug)]
+pub struct ReplicationStats {
```
Comment (Shatur) [inline src/client/diagnostics.rs:13]: How about this name?
```suggestion
pub struct ClientStats {
```

**Review by Shatur — CHANGES_REQUESTED** [review]

File: src/client/diagnostics.rs (hunk 10 above)
Comment (RJ) [reply src/client/diagnostics.rs:13]: yes, better - we may end up with ServerStats later too. will change

**Review by Shatur — APPROVED** [review]

**Comment by Shatur** [comment]
Resolved conflicts.

File: src/client.rs [hunk 11]
```diff
@@ -101,9 +111,14 @@ impl ClientPlugin {
                             &mut entity_map,
                             &replication_rules,
                             tick,
+                            &mut stats,
```
[the remark this answers is not in the document]
Comment (Shatur) [reply src/client.rs]: Thanks, fixed!

**Pull request merged**